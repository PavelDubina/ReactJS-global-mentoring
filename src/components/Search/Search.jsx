import React, { useState } from 'react'
import { useHistory, useParams } from 'react-router-dom'
import { Button } from '../Button/Button'
import { AddForm } from '../AddForm/AddForm'
import { SuccessMessage } from '../SuccessMessage/SuccessMessage'
import styles from './Search.scss'

export const Search = () => {
  const history = useHistory()
  const { query } = useParams()
  const [value, setValue] = useState(query)
  const [{ isForm, isMessage }, setIsOpen] = useState({ isForm: false, isMessage: false })
  const handleChange = (event) => setValue(event.target.value)
  const onOpenModal = () => setIsOpen({ isForm: true, isMessage: false })
  const onCloseModal = (isMessage) => setIsOpen({ isForm: false, isMessage })
  const onSubmit = (event) => {
    if (value) {
      history.push(`/search/${value}`)
    } else {
      history.push(`/`)
    }
    event.preventDefault()
  }
  return (
    <>
      <div className={styles.container}>
        <Button onClick={onOpenModal} styleType="adding">
          + ADD MOVIE
        </Button>
        <h1 className={styles.title}>FIND YOUR MOVIE</h1>
        <form onSubmit={onSubmit} className={styles.form}>
          <input
            onChange={handleChange}
            className={styles.input}
            type="text"
            value={value}
            placeholder="What do you want to watch?"
          />
          <Button type="submit" styleType="search">
            SEARCH
          </Button>
        </form>
      </div>
      {isForm && <AddForm onClose={onCloseModal} />}
      {isMessage && <SuccessMessage onClose={onCloseModal} />}
    </>
  )
}
