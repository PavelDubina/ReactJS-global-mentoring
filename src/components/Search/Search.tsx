import React, { useState, useEffect } from 'react'
import { useRouter } from 'next/router'
import { useTranslation } from 'next-i18next'
import { Button, EStyleTypeBtn, EType } from '../Button/Button'
import { AddForm } from '../AddForm/AddForm'
import { SuccessMessage } from '../SuccessMessage/SuccessMessage'
import styles from '../../../styles/Search.module.scss'

export const Search: React.FC = () => {
  const { t } = useTranslation()
  const router = useRouter()
  const { query } = router.query
  const [value, setValue] = useState('')
  const [{ isForm, isMessage }, setIsOpen] = useState({ isForm: false, isMessage: false })
  const handleChange = (event: React.SyntheticEvent<HTMLInputElement>) => setValue(event.currentTarget.value)
  const onOpenModal = () => setIsOpen({ isForm: true, isMessage: false })
  const onCloseModal = (isMessage?: boolean) => setIsOpen({ isForm: false, isMessage: isMessage ?? false })
  const onSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
    if (value) {
      router.push(`/search/${value}`)
    } else {
      router.push(`/`)
    }
    event.preventDefault()
  }
  useEffect(() => {
    if (router.isReady) setValue(query !== undefined ? String(query) : '')
  }, [query])
  return (
    <>
      <div className={styles.container}>
        <div className={styles.add_btn}>
          <Button onClick={onOpenModal} styleType={EStyleTypeBtn.adding}>
            {t('search.addBtn')}
          </Button>
        </div>
        <h1 className={styles.title}>{t('search.title')}</h1>
        {/* PATTERN:{Controlled input} */}
        <form onSubmit={onSubmit} className={styles.form}>
          <input
            onChange={handleChange}
            className={styles.input}
            type="text"
            value={value}
            placeholder={t('search.placeholder')}
          />
          <div className={styles.search_btn}>
            <Button type={EType.submit} styleType={EStyleTypeBtn.search}>
              {t('search.searchBtn')}
            </Button>
          </div>
        </form>
      </div>
      {isForm && <AddForm onClose={onCloseModal} />}
      {isMessage && <SuccessMessage onClose={onCloseModal} />}
    </>
  )
}
