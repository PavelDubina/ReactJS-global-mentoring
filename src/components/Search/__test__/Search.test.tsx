import React from 'react'
import ReactDOM from 'react-dom'
import { useRouter } from 'next/router'
import 'next-i18next'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { render, fireEvent } from '@testing-library/react'
import { Search } from '../Search'

const mockStore = configureStore([])
jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({ push: jest.fn(), query: { query: '' } }),
}))
jest.mock('next-i18next', () => ({
  useTranslation: jest.fn().mockReturnValue({
    t: (str: any) => str,
  }),
}))
describe('Search', () => {
  ReactDOM.createPortal = jest.fn((element: any) => element)
  const store = mockStore({})
  const getComponent = () =>
    render(
      <Provider store={store}>
        <Search />
      </Provider>,
    )
  it('Should renders correctly', () => {
    expect(getComponent().asFragment()).toMatchSnapshot()
  })
  it('Should open modal after click on button ADD MOVIE and close modal', () => {
    const { getByText } = getComponent()
    const button = getByText('search.addBtn')
    fireEvent.click(button)
    expect(getByText('addModal.title')).toBeTruthy()
    fireEvent.click(getByText('Ã—'))
  })
  it('Should change input value', () => {
    const { getByRole } = getComponent()
    const input: any = getByRole('textbox', { name: '' })
    fireEvent.change(input, { target: { value: 'test value' } })
    expect(input.value).toBe('test value')
  })
  it('Should change pathname after call onSubmit with empty input value', () => {
    const { getByText } = getComponent()
    const submit = getByText('search.searchBtn')
    fireEvent.click(submit)
    expect(useRouter().push).toBeCalled()
  })
  it('Should change pathname after call onSubmit with input value', () => {
    const { getByRole, getByText } = getComponent()
    const submit = getByText('search.searchBtn')
    const input = getByRole('textbox', { name: '' })
    fireEvent.change(input, { target: { value: 'test' } })
    fireEvent.click(submit)
    expect(useRouter().push).toBeCalled()
  })
})
