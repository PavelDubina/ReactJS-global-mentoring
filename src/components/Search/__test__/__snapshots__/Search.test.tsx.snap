// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="add_btn"
    >
      <button
        class="adding"
        type="button"
      >
        search.addBtn
      </button>
    </div>
    <h1
      class="title"
    >
      search.title
    </h1>
    <form
      class="form"
    >
      <input
        class="input"
        placeholder="search.placeholder"
        type="text"
        value=""
      />
      <div
        class="search_btn"
      >
        <button
          class="search"
          type="submit"
        >
          search.searchBtn
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
